File: src/sdk-doc-generators/rust-parser.js
const fs = require('fs');
const path = require('path');

/**
 * Parser for Rust API files to extract method signatures and type information
 */
class RustParser {
    constructor(repoPath, modelsPath) {
        this.repoPath = repoPath;
        this.modelsPath = path.join(repoPath, modelsPath);
        this.typeCache = new Map(); // Cache loaded type definitions
    }

    /**
     * Extract all API methods from a Rust API file
     * @param {string} apiFilePath - Relative path to API file
     * @returns {Array<Object>} - Array of method metadata
     */
    extractApiMethods(apiFilePath) {
        const fullPath = path.join(this.repoPath, apiFilePath);

        if (!fs.existsSync(fullPath)) {
            console.warn(`API file not found: ${fullPath}`);
            return [];
        }

        const content = fs.readFileSync(fullPath, 'utf8');
        const methods = [];

        // Match async function signatures
        // Pattern: pub async fn method_name(configuration: &configuration::Configuration, params: ParamsType) -> Result<ResponseType, Error<ErrorType>>
        const methodRegex = /pub\s+async\s+fn\s+(\w+)\s*\([^,]+,\s*params:\s*(\w+)\)\s*->\s*Result<([^,]+),\s*Error<[^>]+>>/g;

        let match;
        while ((match = methodRegex.exec(content)) !== null) {
            const [, methodName, paramsType, responseType] = match;

            // Extract the method details
            const method = {
                name: methodName,
                paramsType: paramsType.trim(),
                responseType: responseType.trim().replace(/^models::/, ''),
                parameters: {},
                nestedTypes: {}
            };

            // Extract the params struct definition from the same file
            const paramsStruct = this.extractParamsStruct(paramsType, content);
            if (paramsStruct) {
                method.parameters = paramsStruct.fields;

                // Resolve nested types in parameters
                this.resolveNestedTypes(paramsStruct.fields, method.nestedTypes, 0, 3);
            }

            // Resolve the response type (use method.responseType which has models:: stripped)
            const responseTypeDef = this.loadTypeDefinition(method.responseType);

            if (responseTypeDef) {
                method.nestedTypes[method.responseType] = {
                    summary: responseTypeDef.summary,
                    filePath: responseTypeDef.filePath
                };

                // Resolve nested types in response (shallower depth to avoid bloat)
                if (responseTypeDef.fields) {
                    this.resolveNestedTypes(responseTypeDef.fields, method.nestedTypes, 0, 2);
                }
            }

            methods.push(method);
        }

        return methods;
    }

    /**
     * Extract parameter struct definition from content
     * @param {string} structName - Struct name to find
     * @param {string} content - File content to search in
     * @returns {Object|null} - Struct metadata
     */
    extractParamsStruct(structName, content) {
        // Match: pub struct StructName { ... }
        // Need to handle nested braces properly
        const structStartRegex = new RegExp(
            `pub\\s+struct\\s+${structName}\\s*\\{`,
            ''
        );

        const startMatch = content.match(structStartRegex);
        if (!startMatch) {
            return null;
        }

        // Find the matching closing brace
        const startIndex = startMatch.index + startMatch[0].length;
        let braceCount = 1;
        let endIndex = startIndex;

        for (let i = startIndex; i < content.length; i++) {
            if (content[i] === '{') braceCount++;
            if (content[i] === '}') braceCount--;

            if (braceCount === 0) {
                endIndex = i;
                break;
            }
        }

        const fieldsText = content.substring(startIndex, endIndex);
        const fields = {};

        // Match fields: pub field_name: Type
        // Handle Option<Type> for optional fields
        const fieldRegex = /pub\s+(\w+):\s*([^,\n]+)/g;

        let fieldMatch;
        while ((fieldMatch = fieldRegex.exec(fieldsText)) !== null) {
            const [, fieldName, fieldType] = fieldMatch;
            const cleanType = fieldType.trim();

            // Check if it's an Option<T> type
            const isOptional = cleanType.startsWith('Option<');
            let actualType = cleanType;
            if (isOptional) {
                const optionMatch = cleanType.match(/Option<(.+)>/);
                if (optionMatch) {
                    actualType = optionMatch[1];
                }
            }

            fields[fieldName] = {
                type: actualType,
                required: !isOptional
            };
        }

        return {
            name: structName,
            fields
        };
    }

    /**
     * Resolve nested type definitions
     * @param {Object} fields - Fields to resolve
     * @param {Object} nestedTypes - Output object for nested types
     * @param {number} currentDepth - Current recursion depth
     * @param {number} maxDepth - Maximum recursion depth
     */
    resolveNestedTypes(fields, nestedTypes, currentDepth, maxDepth) {
        if (currentDepth >= maxDepth) {
            return;
        }

        for (const [fieldName, fieldInfo] of Object.entries(fields)) {
            const type = fieldInfo.type;

            // Extract type names (handle Vec, Option, etc.)
            const typeNames = this.extractTypeNames(type);

            for (const typeName of typeNames) {
                // Skip primitive types and already resolved types
                if (this.isPrimitiveType(typeName) || nestedTypes[typeName]) {
                    continue;
                }

                // Try to load type definition from models
                const typeDef = this.loadTypeDefinition(typeName);
                if (typeDef) {
                    nestedTypes[typeName] = {
                        summary: typeDef.summary,
                        filePath: typeDef.filePath
                    };

                    // Recursively resolve nested types
                    if (typeDef.fields) {
                        this.resolveNestedTypes(typeDef.fields, nestedTypes, currentDepth + 1, maxDepth);
                    }
                }
            }
        }
    }

    /**
     * Extract type names from a type expression
     * @param {string} typeExpr - Type expression (e.g., "Vec<Type>", "Box<Type>")
     * @returns {Array<string>} - Array of type names
     */
    extractTypeNames(typeExpr) {
        const names = [];

        // Remove common wrappers and clean up
        typeExpr = typeExpr.trim();

        // Handle Vec<Type>
        const vecMatch = typeExpr.match(/Vec<(.+)>/);
        if (vecMatch) {
            names.push(...this.extractTypeNames(vecMatch[1]));
            return names;
        }

        // Handle Box<Type>
        const boxMatch = typeExpr.match(/Box<(.+)>/);
        if (boxMatch) {
            names.push(...this.extractTypeNames(boxMatch[1]));
            return names;
        }

        // Handle std::collections::HashMap<K, V> - extract V
        const hashMapMatch = typeExpr.match(/std::collections::HashMap<[^,]+,\s*(.+)>/);
        if (hashMapMatch) {
            names.push(...this.extractTypeNames(hashMapMatch[1]));
            return names;
        }

        // Remove models:: prefix if present
        typeExpr = typeExpr.replace(/^models::/, '');

        // Simple type name
        if (!this.isPrimitiveType(typeExpr)) {
            names.push(typeExpr);
        }

        return names;
    }

    /**
     * Check if type is a primitive
     * @param {string} typeName - Type name
     * @returns {boolean}
     */
    isPrimitiveType(typeName) {
        const primitives = ['String', 'str', 'i32', 'i64', 'u32', 'u64', 'f32', 'f64', 'bool', 'usize', 'isize'];
        return primitives.includes(typeName);
    }

    /**
     * Load type definition from models directory
     * @param {string} typeName - Type name to load
     * @returns {Object|null} - Type definition
     */
    loadTypeDefinition(typeName) {
        // Check cache first
        if (this.typeCache.has(typeName)) {
            return this.typeCache.get(typeName);
        }

        // Convert CamelCase to snake_case for file names
        const fileName = this.toSnakeCase(typeName);
        const typeFile = path.join(this.modelsPath, `${fileName}.rs`);

        if (!fs.existsSync(typeFile)) {
            return null;
        }

        const content = fs.readFileSync(typeFile, 'utf8');

        // Check if it's an enum
        const enumDef = this.extractEnum(typeName, content);
        if (enumDef) {
            enumDef.filePath = `${fileName}.rs`;
            this.typeCache.set(typeName, enumDef);
            return enumDef;
        }

        // Extract struct
        const structDef = this.extractStruct(typeName, content);
        if (structDef) {
            const typeDef = {
                name: typeName,
                kind: 'struct',
                fields: structDef.fields,
                summary: this.summarizeStruct(structDef),
                filePath: `${fileName}.rs`
            };
            this.typeCache.set(typeName, typeDef);
            return typeDef;
        }

        return null;
    }

    /**
     * Extract struct definition from content
     * @param {string} structName - Struct name to find
     * @param {string} content - File content
     * @returns {Object|null} - Struct metadata
     */
    extractStruct(structName, content) {
        // Match: pub struct StructName { ... }
        const structStartRegex = new RegExp(
            `pub\\s+struct\\s+${structName}\\s*\\{`,
            ''
        );

        const startMatch = content.match(structStartRegex);
        if (!startMatch) {
            return null;
        }

        // Find the matching closing brace
        const startIndex = startMatch.index + startMatch[0].length;
        let braceCount = 1;
        let endIndex = startIndex;

        for (let i = startIndex; i < content.length; i++) {
            if (content[i] === '{') braceCount++;
            if (content[i] === '}') braceCount--;

            if (braceCount === 0) {
                endIndex = i;
                break;
            }
        }

        const fieldsText = content.substring(startIndex, endIndex);
        const fields = {};

        // Match fields with serde annotations
        // #[serde(...)] pub field_name: Type,
        const lines = fieldsText.split('\n');
        let currentField = null;
        let isOptional = false;

        for (const line of lines) {
            // Check for skip_serializing_if = "Option::is_none" which indicates optional
            if (line.includes('skip_serializing_if = "Option::is_none"')) {
                isOptional = true;
            }

            // Match field definitions
            const fieldMatch = line.match(/pub\s+(\w+):\s*([^,]+)/);
            if (fieldMatch) {
                const [, fieldName, fieldType] = fieldMatch;
                const cleanType = fieldType.trim();

                // Check if it's an Option<T> type
                const typeIsOption = cleanType.startsWith('Option<');
                let actualType = cleanType;
                if (typeIsOption) {
                    const optionMatch = cleanType.match(/Option<(.+)>/);
                    if (optionMatch) {
                        actualType = optionMatch[1];
                    }
                }

                fields[fieldName] = {
                    type: actualType,
                    required: !(isOptional || typeIsOption)
                };

                isOptional = false; // Reset for next field
            }
        }

        return {
            name: structName,
            fields
        };
    }

    /**
     * Extract enum definition from content
     * @param {string} enumName - Enum name to find
     * @param {string} content - File content
     * @returns {Object|null} - Enum metadata
     */
    extractEnum(enumName, content) {
        // Match: pub enum EnumName { ... }
        const enumRegex = new RegExp(
            `pub\\s+enum\\s+${enumName}\\s*\\{([^}]+)\\}`,
            's'
        );

        const match = content.match(enumRegex);
        if (!match) {
            return null;
        }

        const variantsText = match[1];
        const variants = [];

        // Match enum variants - handle both simple and complex variants
        // Simple: VariantName,
        // With data: VariantName(Type),
        const lines = variantsText.split('\n');
        for (const line of lines) {
            const simpleMatch = line.match(/#\[serde\(rename\s*=\s*"([^"]+)"\)\]/);
            if (simpleMatch) {
                variants.push(simpleMatch[1]);
            }
        }

        // If no serde renames found, try simple variant names
        if (variants.length === 0) {
            const variantRegex = /(\w+)(?:\([^)]*\))?\s*,/g;
            let variantMatch;
            while ((variantMatch = variantRegex.exec(variantsText)) !== null) {
                variants.push(this.toSnakeCase(variantMatch[1]));
            }
        }

        return {
            name: enumName,
            kind: 'enum',
            variants,
            summary: `enum: ${variants.map(v => `'${v}'`).join(' | ')}`
        };
    }

    /**
     * Summarize struct as a string
     * @param {Object} structDef - Struct definition
     * @returns {string} - Summary string
     */
    summarizeStruct(structDef) {
        const fields = [];
        for (const [name, info] of Object.entries(structDef.fields)) {
            const optional = info.required ? '' : '?';
            fields.push(`${name}${optional}: ${info.type}`);
        }
        return `{ ${fields.join(', ')} }`;
    }

    /**
     * Convert CamelCase to snake_case
     * @param {string} str - CamelCase string
     * @returns {string} - snake_case string
     */
    toSnakeCase(str) {
        return str
            .replace(/([A-Z])/g, '_$1')
            .replace(/([0-9]+)/g, '_$1')
            .toLowerCase()
            .replace(/^_/, '');
    }
}

module.exports = RustParser;

File: rust/sdkgen/src/generators/ai/rust_parser.rs
//! Port of `src/sdk-doc-generators/rust-parser.js` (441 LOC).
//!
//! Parses Rust API source files, extracting `pub async fn` signatures
//! and resolving param-struct fields + nested model types from
//! `client/src/models/`.

use std::cell::RefCell;
use std::collections::BTreeMap;
use std::path::PathBuf;

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};

use super::typescript_parser::NestedType;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParamInfo {
    #[serde(rename = "type")]
    pub type_: String,
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Method {
    pub name: String,
    #[serde(rename = "paramsType")]
    pub params_type: String,
    #[serde(rename = "responseType")]
    pub response_type: String,
    #[serde(serialize_with = "serialize_indexmap")]
    pub parameters: indexmap::IndexMap<String, ParamInfo>,
    #[serde(rename = "nestedTypes", serialize_with = "serialize_indexmap")]
    pub nested_types: indexmap::IndexMap<String, NestedType>,
    #[serde(rename = "httpMethod", default, skip_serializing_if = "Option::is_none")]
    pub http_method: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tag: Option<String>,
    #[serde(rename = "authType", default, skip_serializing_if = "Option::is_none")]
    pub auth_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

fn serialize_indexmap<S, K, V>(
    map: &indexmap::IndexMap<K, V>,
    ser: S,
) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
    K: serde::Serialize + Eq + std::hash::Hash,
    V: serde::Serialize,
{
    use serde::ser::SerializeMap;
    let mut m = ser.serialize_map(Some(map.len()))?;
    for (k, v) in map {
        m.serialize_entry(k, v)?;
    }
    m.end()
}

pub struct RustParser {
    pub repo_path: PathBuf,
    pub models_path: PathBuf,
    type_cache: RefCell<BTreeMap<String, Option<TypeDef>>>,
}

#[derive(Debug, Clone)]
struct TypeDef {
    #[allow(dead_code)]
    name: String,
    kind: &'static str, // "struct" or "enum"
    fields: indexmap::IndexMap<String, ParamInfo>,
    summary: String,
    file_path: String,
}

impl RustParser {
    pub fn new(repo_path: PathBuf, models_path_rel: &str) -> Self {
        let models_path = repo_path.join(models_path_rel);
        Self {
            repo_path,
            models_path,
            type_cache: RefCell::new(BTreeMap::new()),
        }
    }

    pub fn extract_api_methods(&self, api_file_rel: &str) -> Vec<Method> {
        let path = self.repo_path.join(api_file_rel);
        let Ok(content) = std::fs::read_to_string(&path) else {
            tracing::warn!(path = %path.display(), "rust API file not found");
            return Vec::new();
        };
        let mut out = Vec::new();
        // Mirrors rust-parser.js:32.
        static SIG_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(
                r"pub\s+async\s+fn\s+(\w+)\s*\([^,]+,\s*params:\s*(\w+)\)\s*->\s*Result<([^,]+),\s*Error<[^>]+>>",
            )
            .expect("regex")
        });
        for cap in SIG_RE.captures_iter(&content) {
            let name = cap[1].to_string();
            let params_type = cap[2].trim().to_string();
            let mut response_type = cap[3].trim().to_string();
            if let Some(stripped) = response_type.strip_prefix("models::") {
                response_type = stripped.to_string();
            }

            let mut method = Method {
                name,
                params_type: params_type.clone(),
                response_type: response_type.clone(),
                parameters: indexmap::IndexMap::new(),
                nested_types: indexmap::IndexMap::new(),
                http_method: None,
                path: None,
                tag: None,
                auth_type: None,
                description: None,
            };
            // Mirror Node's two-function shape: parameters use
            // extractParamsStruct (pure type-based), nested types use
            // extractStruct (combined check). Using the combined-check
            // extractor for both — as we did before — diverged from
            // Node on any params struct whose fields had a
            // `#[serde(skip_serializing_if = "Option::is_none")]`
            // attribute. Different `required` flags propagate into the
            // cache key + the prompt, so the next cache miss
            // re-generates with a subtly different signature.
            if let Some(s) = self.extract_params_struct(&params_type, &content) {
                method.parameters = s.fields.clone();
                self.resolve_nested_types(&s.fields, &mut method.nested_types, 0, 3);
            }
            if let Some(td) = self.load_type_definition(&response_type) {
                method.nested_types.insert(
                    response_type.clone(),
                    NestedType {
                        summary: td.summary.clone(),
                        file_path: td.file_path.clone(),
                    },
                );
                if td.kind == "struct" {
                    self.resolve_nested_types(&td.fields, &mut method.nested_types, 0, 2);
                }
            }
            out.push(method);
        }
        out
    }

    /// Extract a method's params struct using Node's
    /// `extractParamsStruct` semantics (line 83-142 of rust-parser.js):
    /// `required` is set purely from whether the field type starts
    /// with `Option<`. Serde attributes are NOT consulted.
    ///
    /// This differs from `extract_struct` (which mirrors Node
    /// `extractStruct` at line 291-362) by NOT carrying forward
    /// `#[serde(skip_serializing_if = "Option::is_none")]` from a
    /// preceding line. Using extract_struct for parameters silently
    /// diverged from Node on any params struct whose fields were
    /// annotated, which would have produced different cache hashes
    /// and prompts on regeneration.
    fn extract_params_struct(&self, name: &str, content: &str) -> Option<StructDef> {
        let body = extract_struct_body(name, content)?;
        // Node uses /pub\s+(\w+):\s*([^,\n]+)/g — note the \n in the
        // negation, so each field is captured per-line. Mirror that
        // exactly so multi-line declarations behave the same as Node.
        static FIELD_RE: Lazy<Regex> =
            Lazy::new(|| Regex::new(r"pub\s+(\w+):\s*([^,\n]+)").expect("regex"));
        let mut fields: indexmap::IndexMap<String, ParamInfo> = indexmap::IndexMap::new();
        for cap in FIELD_RE.captures_iter(&body) {
            let field_name = cap[1].to_string();
            let clean = cap[2].trim().to_string();
            let type_is_option = clean.starts_with("Option<");
            let actual = if type_is_option {
                Regex::new(r"Option<(.+)>")
                    .ok()
                    .and_then(|r| r.captures(&clean))
                    .map(|c| c[1].to_string())
                    .unwrap_or_else(|| clean.clone())
            } else {
                clean
            };
            fields.insert(
                field_name,
                ParamInfo {
                    type_: actual,
                    // Pure type-based; no skip_serializing_if check.
                    required: !type_is_option,
                },
            );
        }
        Some(StructDef {
            name: name.to_string(),
            fields,
        })
    }

    fn extract_struct(&self, name: &str, content: &str) -> Option<StructDef> {
        let body = extract_struct_body(name, content)?;
        let mut fields: indexmap::IndexMap<String, ParamInfo> = indexmap::IndexMap::new();
        let mut is_optional = false;
        for line in body.split('\n') {
            if line.contains("skip_serializing_if = \"Option::is_none\"") {
                is_optional = true;
            }
            static FIELD_RE: Lazy<Regex> =
                Lazy::new(|| Regex::new(r"pub\s+(\w+):\s*([^,]+)").expect("regex"));
            if let Some(cap) = FIELD_RE.captures(line) {
                let name = cap[1].to_string();
                let clean = cap[2].trim().to_string();
                let type_is_option = clean.starts_with("Option<");
                let actual = if type_is_option {
                    Regex::new(r"Option<(.+)>")
                        .ok()
                        .and_then(|r| r.captures(&clean))
                        .map(|c| c[1].to_string())
                        .unwrap_or(clean.clone())
                } else {
                    clean
                };
                fields.insert(
                    name,
                    ParamInfo {
                        type_: actual,
                        required: !(is_optional || type_is_option),
                    },
                );
                is_optional = false;
            }
        }
        Some(StructDef {
            name: name.to_string(),
            fields,
        })
    }

    fn resolve_nested_types(
        &self,
        fields: &indexmap::IndexMap<String, ParamInfo>,
        out: &mut indexmap::IndexMap<String, NestedType>,
        current: usize,
        max: usize,
    ) {
        if current >= max {
            return;
        }
        for info in fields.values() {
            for type_name in self.extract_type_names(&info.type_) {
                if Self::is_primitive(&type_name) || out.contains_key(&type_name) {
                    continue;
                }
                if let Some(td) = self.load_type_definition(&type_name) {
                    out.insert(
                        type_name.clone(),
                        NestedType {
                            summary: td.summary.clone(),
                            file_path: td.file_path.clone(),
                        },
                    );
                    if td.kind == "struct" {
                        self.resolve_nested_types(&td.fields, out, current + 1, max);
                    }
                }
            }
        }
    }

    fn extract_type_names(&self, expr: &str) -> Vec<String> {
        let s = expr.trim();
        // Vec<T>
        if let Some(c) = Regex::new(r"Vec<(.+)>").ok().and_then(|r| r.captures(s).map(|c| c[1].to_string())) {
            return self.extract_type_names(&c);
        }
        // Box<T>
        if let Some(c) = Regex::new(r"Box<(.+)>").ok().and_then(|r| r.captures(s).map(|c| c[1].to_string())) {
            return self.extract_type_names(&c);
        }
        // HashMap<K, V>
        if let Some(c) = Regex::new(r"std::collections::HashMap<[^,]+,\s*(.+)>")
            .ok()
            .and_then(|r| r.captures(s).map(|c| c[1].to_string()))
        {
            return self.extract_type_names(&c);
        }
        let stripped = s.strip_prefix("models::").unwrap_or(s).to_string();
        if Self::is_primitive(&stripped) {
            return Vec::new();
        }
        vec![stripped]
    }

    fn is_primitive(t: &str) -> bool {
        matches!(
            t,
            "String" | "str" | "i32" | "i64" | "u32" | "u64" | "f32" | "f64" | "bool" | "usize" | "isize"
        )
    }

    fn load_type_definition(&self, type_name: &str) -> Option<TypeDef> {
        if let Some(cached) = self.type_cache.borrow().get(type_name) {
            return cached.clone();
        }
        let file_name = Self::to_snake_case(type_name);
        let type_file = self.models_path.join(format!("{file_name}.rs"));
        if !type_file.exists() {
            self.type_cache
                .borrow_mut()
                .insert(type_name.to_string(), None);
            return None;
        }
        let Ok(content) = std::fs::read_to_string(&type_file) else {
            self.type_cache
                .borrow_mut()
                .insert(type_name.to_string(), None);
            return None;
        };
        // Enum?
        if let Some(variants) = self.extract_enum(type_name, &content) {
            let summary = format!(
                "enum: {}",
                variants
                    .iter()
                    .map(|v| format!("'{v}'"))
                    .collect::<Vec<_>>()
                    .join(" | ")
            );
            let td = TypeDef {
                name: type_name.to_string(),
                kind: "enum",
                fields: indexmap::IndexMap::new(),
                summary,
                file_path: format!("{file_name}.rs"),
            };
            self.type_cache
                .borrow_mut()
                .insert(type_name.to_string(), Some(td.clone()));
            return Some(td);
        }
        // Struct?
        if let Some(s) = self.extract_struct(type_name, &content) {
            let summary = Self::summarize_struct(&s);
            let td = TypeDef {
                name: type_name.to_string(),
                kind: "struct",
                fields: s.fields,
                summary,
                file_path: format!("{file_name}.rs"),
            };
            self.type_cache
                .borrow_mut()
                .insert(type_name.to_string(), Some(td.clone()));
            return Some(td);
        }
        self.type_cache
            .borrow_mut()
            .insert(type_name.to_string(), None);
        None
    }

    fn extract_enum(&self, name: &str, content: &str) -> Option<Vec<String>> {
        let re = Regex::new(&format!(
            r"(?s)pub\s+enum\s+{}\s*\{{([^}}]+)\}}",
            regex::escape(name)
        ))
        .ok()?;
        let cap = re.captures(content)?;
        let body = &cap[1];
        let mut variants = Vec::new();
        for line in body.split('\n') {
            if let Some(c) =
                Regex::new(r#"#\[serde\(rename\s*=\s*"([^"]+)"\)\]"#)
                    .ok()
                    .and_then(|r| r.captures(line).map(|c| c[1].to_string()))
            {
                variants.push(c);
            }
        }
        if variants.is_empty() {
            // Fallback: extract bare variant names (snake_case via to_snake_case).
            let re = Regex::new(r"(\w+)(?:\([^)]*\))?\s*,").expect("regex");
            for c in re.captures_iter(body) {
                variants.push(Self::to_snake_case(&c[1]));
            }
        }
        Some(variants)
    }

    fn summarize_struct(s: &StructDef) -> String {
        let mut parts = Vec::new();
        for (name, info) in &s.fields {
            let opt = if info.required { "" } else { "?" };
            parts.push(format!("{name}{opt}: {}", info.type_));
        }
        format!("{{ {} }}", parts.join(", "))
    }

    /// Mirrors `rust-parser.js:432-438` — CamelCase -> snake_case
    /// with a digit-insertion rule.
    pub fn to_snake_case(s: &str) -> String {
        // Step 1: insert `_` before each uppercase letter.
        let mut out = String::new();
        for c in s.chars() {
            if c.is_ascii_uppercase() {
                out.push('_');
            }
            out.push(c);
        }
        // Step 2: insert `_` before each digit run. JS `.replace(/([0-9]+)/g, '_$1')`.
        let with_digits = Regex::new(r"([0-9]+)").unwrap().replace_all(&out, "_$1").into_owned();
        let lower = with_digits.to_lowercase();
        lower.trim_start_matches('_').to_string()
    }
}

struct StructDef {
    #[allow(dead_code)]
    name: String,
    fields: indexmap::IndexMap<String, ParamInfo>,
}

/// Locate `pub struct <name> { ... }` in `content` and return the
/// brace-balanced body. Shared by both `extract_params_struct` and
/// `extract_struct` — they only differ in how they interpret the
/// body, not in how they slice it.
fn extract_struct_body<'a>(name: &str, content: &'a str) -> Option<&'a str> {
    let re = Regex::new(&format!(
        r"pub\s+struct\s+{}\s*\{{",
        regex::escape(name)
    ))
    .ok()?;
    let m = re.find(content)?;
    let start = m.end();
    let bytes = content.as_bytes();
    let mut depth = 1i32;
    let mut end = start;
    let mut i = start;
    while i < bytes.len() {
        match bytes[i] {
            b'{' => depth += 1,
            b'}' => {
                depth -= 1;
                if depth == 0 {
                    end = i;
                    break;
                }
            }
            _ => {}
        }
        i += 1;
    }
    Some(&content[start..end])
}

/// snake_case -> camelCase with acronym handling for OpenAPI name
/// matching. Mirrors `rust-ai-generator.js:357-401`.
pub fn snake_to_camel_case(s: &str) -> String {
    const ACRONYMS: &[&str] = &["sso", "api", "url", "uri", "http", "https", "html", "json", "xml"];
    let parts: Vec<&str> = s.split('_').collect();
    let mut out = String::new();
    for (i, part) in parts.iter().enumerate() {
        let lower = part.to_lowercase();
        if lower == "id" {
            out.push_str("Id");
            continue;
        }
        if ACRONYMS.contains(&lower.as_str()) {
            out.push_str(&part.to_uppercase());
            continue;
        }
        // Compound acronym at suffix.
        let mut handled = false;
        for ac in ACRONYMS {
            if lower.ends_with(ac) && lower != *ac {
                let prefix = &lower[..lower.len() - ac.len()];
                if ACRONYMS.contains(&prefix) {
                    out.push_str(&prefix.to_uppercase());
                    let mut chars = ac.chars();
                    if let Some(first) = chars.next() {
                        out.extend(first.to_uppercase());
                        out.push_str(chars.as_str());
                    }
                    handled = true;
                    break;
                }
            }
        }
        if handled {
            continue;
        }
        if lower.ends_with("id") && lower != "id" {
            let prefix = &lower[..lower.len() - 2];
            if ACRONYMS.contains(&prefix) {
                out.push_str(&prefix.to_uppercase());
                out.push_str("Id");
                continue;
            }
        }
        if i == 0 {
            out.push_str(part);
        } else {
            let mut chars = part.chars();
            if let Some(first) = chars.next() {
                out.extend(first.to_uppercase());
                out.push_str(chars.as_str());
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    //! Pins the params-vs-nested-type divergence Node has at
    //! src/sdk-doc-generators/rust-parser.js — extractParamsStruct
    //! (line 83-142) is pure type-based; extractStruct (line 291-362)
    //! also carries forward `skip_serializing_if = "Option::is_none"`
    //! from the preceding line. The Rust port used the combined
    //! extractor for BOTH, so params silently picked up the carry-
    //! forward behavior and produced different `required` flags than
    //! Node — invalidating cache hits and changing prompt content.
    use super::*;
    use std::path::PathBuf;

    fn parser() -> RustParser {
        // models_path doesn't matter for the in-memory body tests
        // below; load_type_definition is what touches disk.
        RustParser::new(PathBuf::from("/tmp"), "client/src/models/")
    }

    #[test]
    fn params_struct_ignores_serde_skip_attribute() {
        // The exact divergence shape Node + Rust differ on:
        // a `#[serde(skip_serializing_if = "Option::is_none")]` line
        // followed by a NON-Option field. Node extractParamsStruct
        // ignores serde attrs entirely -> required = true. Old Rust
        // extract_struct carried the attr forward -> required = false.
        let source = r#"
pub struct AddDomainConfigParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tenant_id: String,
    pub add_domain_config_params: models::AddDomainConfigParams
}
"#;
        let s = parser()
            .extract_params_struct("AddDomainConfigParams", source)
            .expect("struct present");
        let tenant = s.fields.get("tenant_id").expect("tenant_id field");
        assert!(
            tenant.required,
            "params extractor must ignore #[serde(skip_serializing_if=…)]; \
             Node marks this field required=true (line 83-142 pure type-based)"
        );
    }

    #[test]
    fn params_struct_marks_option_fields_optional() {
        // Normal case (which Node + the old combined extractor agreed
        // on) — keep it locked so refactors don't drift back.
        let source = r#"
pub struct GetTicketsParams {
    pub tenant_id: Option<String>,
    pub status: String,
}
"#;
        let s = parser()
            .extract_params_struct("GetTicketsParams", source)
            .expect("struct present");
        assert_eq!(
            s.fields.get("tenant_id").map(|p| p.required),
            Some(false),
            "Option<T> -> required=false"
        );
        assert_eq!(
            s.fields.get("status").map(|p| p.required),
            Some(true),
            "non-Option -> required=true"
        );
    }

    #[test]
    fn params_unwraps_option_inner_type() {
        let source = r#"
pub struct P {
    pub a: Option<MyType>,
    pub b: Option<Vec<u8>>,
}
"#;
        let s = parser().extract_params_struct("P", source).unwrap();
        assert_eq!(s.fields.get("a").map(|p| p.type_.as_str()), Some("MyType"));
        assert_eq!(s.fields.get("b").map(|p| p.type_.as_str()), Some("Vec<u8>"));
    }

    /// extract_struct (used for nested types) is the one that DOES
    /// match Node's combined check at extractStruct line 291-362.
    /// Pin that semantics so the split doesn't accidentally flip
    /// the nested-type extractor too.
    #[test]
    fn nested_struct_extractor_keeps_combined_check() {
        let source = r#"
pub struct ResponseModel {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub maybe_field: Option<String>,
    pub required_field: String,
}
"#;
        let s = parser()
            .extract_struct("ResponseModel", source)
            .expect("struct present");
        // maybe_field: both attr AND Option<...> -> required=false
        assert_eq!(
            s.fields.get("maybe_field").map(|p| p.required),
            Some(false)
        );
        // required_field: no attr, no Option -> required=true
        assert_eq!(
            s.fields.get("required_field").map(|p| p.required),
            Some(true)
        );
    }
}
